### DISCUSSION TITLE
Exception in checkout_updated hook should rollback the checkout change

### DISCUSSION THREAD

**Discussion by satvikt** [saleor/saleor#9488] [General]
I faced a situation wherein I added an item to checkout, and encountered an exception in the `checkout_updated` hook of my custom plugin. 
I expected the item to have not been put into checkout (or for it to have been removed from checkout after  the exception occurred). But the item was added to the checkout nonetheless.

I observed that such a behaviour can be prevented by placing the code block under a `traced_atomic_transaction` context manager, which commits the transaction only if no exception is encountered in the entire block. Otherwise, it rolls back the transaction.

However, this context manager - traced_atomic_transaction - has been sparingly used in mutations,
E.g. In the `CheckoutShippingAddressUpdate`

```
with traced_atomic_transaction():
      shipping_address.save()
      change_shipping_address_in_checkout(
          checkout_info, shipping_address, lines, discounts, manager
      )
```
likewise, in CheckoutComplete.

I have 2 queries here:
1. Is there any reason as to why the "perform_mutation" code of checkout mutations was not kept under an atomic transaction? So that any exception under the mutation causes the entire mutation to roll back?

In other words, is the non-atomicity of mutations something intentional?

2. In case I wish to achieve such behaviour - an exception raised under `checkout_updated` hook in my plugin should cause the mutation to roll back - what's the best way to do this, without modifying the code written in saleor-core?

PS: I have tried the above with Saleor Core 3.0.7. Shall update the thread once I try it out with the latest 3.1 release.

**Comment by karolkielecki** [comment]
Hi @satvikt, let me ask the team. Please ping me if I don't get back to you this week.

**Comment by satvikt** [comment]
<Update>

Confirmed that the latest 3.1 release has the same behaviour.

**Comment by korycins** [comment]
Hey, we didn't treat these hooks as something that allows you to handle an additional validation. Most of these hooks should be placed inside the transaction.on_commit.
We're working on a fully new checkout flow, that should reduce the complexity of the process. Here you can find some charts to give you more context: https://github.com/saleor/saleor/pull/9343. 
The assumption is that we will have a separate checkout app that will handle a checkout process, including a payment (and could include your custom validation) and calling Saleor to create an order.  It is still WIP. Here you can track our sample checkout app: https://github.com/saleor/saleor-checkout
Because of this, we're thinking about extracting logic from WebhookPlugin to Saleor and deprecating a plugin's interface as all custom actions should be done by separate Saleor Apps.

If you still need a way to handle a custom validation, I would check & consider `perform_mutation` hook.
